src/Interpreters/Context.cpp

```diff
@@ -1644,6 +1644,12 @@ void Context::addQueryAccessInfo(const QualifiedProjectionName & qualified_proje
         "{}.{}", qualified_projection_name.storage_id.getFullTableName(), backQuoteIfNeed(qualified_projection_name.projection_name)));
 }
 
+Context::QueryFactoriesInfo Context::getQueryFactoriesInfo() const
+{
+    std::lock_guard lock(query_factories_info.mutex);
```

Comment:
1It seems this lock is not necessary. The lock is taken in copy constructor on the right-hand object. See https://github.com/ClickHouse/ClickHouse/blob/9adab57771c75f691d4dec73fd3abc88d516206f/src/Interpreters/Context.h#L362 

Also, it seems assignment operator and swap() method are not in use - https://github.com/ClickHouse/ClickHouse/blob/9adab57771c75f691d4dec73fd3abc88d516206f/src/Interpreters/Context.h#L376-L393 . These methods don't take the lock. So, we either fix them, or better delete them.